### PR TITLE
Transport Rewrite

### PR DESCRIPTION
**Pull request by James-Frowen**
### Todo:
- [ ] multithread support
- [ ] validate new connections
  - [ ] key/phrase
  - [ ] hashcash
- [ ] decide on Key for Connection (replaces `EndPoint`)

### This needs to split into smaller PR:
- [x] Rename NetworkConnection To NetworkPlayer (https://github.com/MirageNet/Mirage/pull/684)
  - [ ] Update uses of NetworkPlayer 
- [ ] Remove INetworkPlayer, use smaller interfaces instead
  - [x] Split INetworkPlayer
- [ ] Remove Nofity from NetworkConnection
- [ ] Replace Transport with Peer
- [ ] Replacce ObjectManagers with World
- [ ] Remove Channels (use bool instead?)

### Resources:
- https://github.com/fholm/StreamTransport/commit/b1279cbda85c8aa6759493b3c3a074b014cec7ea
- https://github.com/phodoval/Mirage/commit/7db8f5128119499505b4783511d0b174cfb2dde6

### Concepts
World
- List of NetworkIdentities
- List of Player
- Handles message

InterestManagerment (AOI)
- uses List of NetworkIdentities and List of Player
- decides which player can see which Identity

Player
- owns NetworkIdentities
- views NetworkIdentities (controled by AOI/Visbility)
- has a connection (optional)

Peer 
- gate keeper
- has list of connections

ISocket
- gate to outside World

connections
- has list of players
- can send and receive data (raw byte)
    - can be reliable/unreliable/notify

messages
- unpacked/deserialized form of raw byte
- has message Id
- messages are sent between Machines, no sending remote message to local handlers

Machines
- unity instance 

RPC
- function that exists on GameObjects
- Rpc target players 
- if player has no connection, invoke locally
- if player has connection, invoke remotely
- if "send to all" repeat above for each player

### CODE REVIEW

File: Assets/Mirage/Runtime/Transport/ISocket.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        /// </summary>
+        /// <returns>true if data to read</returns>
+        bool Poll();
+
+        /// <summary>
+        /// Gets next Message
+        /// <para>Should be called after Poll</para>
+        /// </summary>
+        /// <param name="data">recieved data</param>
+        void Recieve(byte[] data, ref EndPoint endPoint, out int bytesReceived);
```
Comment (paulpach) [inline Assets/Mirage/Runtime/Transport/ISocket.cs:37]:
```suggestion
        int Recieve(byte[] data, ref EndPoint endPoint);
```
no out parameters please

File: Assets/Mirage/Runtime/Transport/SocketLayer/ISocket.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        /// </summary>
+        /// <returns>true if data to read</returns>
+        bool Poll();
+
+        /// <summary>
+        /// Gets next Message
+        /// <para>Should be called after Poll</para>
+        /// </summary>
+        /// <param name="data">recieved data</param>
+        void Recieve(byte[] data, ref EndPoint endPoint, out int bytesReceived);
```
Comment (Kerozard) [inline Assets/Mirage/Runtime/Transport/SocketLayer/ISocket.cs:30]: This seems to be a typo. It should be Receive.

Comment (James-Frowen) [reply Assets/Mirage/Runtime/Transport/SocketLayer/ISocket.cs:30]: Thanks for spotting this, I'll make sure fix this soon

**Comment by James-Frowen** [comment]
new branch https://github.com/MirageNet/Mirage/pull/749

**Pull request closed without merging**